### PR TITLE
Query protocol wiring

### PR DESCRIPTION
**Pull request by Baccata**
Supersedes https://github.com/disneystreaming/smithy4s/pull/723 

Partially solves https://github.com/disneystreaming/smithy4s/issues/720

@igor-ramazanov, could you try this branch ? I've added [this bit](https://github.com/disneystreaming/smithy4s/compare/feature/aws-query-protocol...query-protocol-wiring-baccata?expand=1#diff-33e6cdc4dccd29147741c838c0969f398bb02072d9bfc1986b77c99f20415f12R137). 

it occurs to me that there was a simpler/cleaner path than amending the endpoint by deferring the instantiation of the codecAPI to where the information was available. Sorry for giving you the wrong information earlier

### CODE REVIEW

File: modules/aws/src/smithy4s/aws/internals/AwsQueryRPCInterpreter.scala [hunk 1]
```diff
@@ -75,8 +56,10 @@ private[aws] class AwsQueryRPCInterpreter[Alg[_[_, _, _, _, _]], Op[_,_,_,_,_],
     ] {
       def apply[I, E, O, SI, SO](
           endpoint: Endpoint[Op, I, E, O, SI, SO]
-      ): AwsUnaryEndpoint[F, Op, I, E, O, SI, SO] =
-        new AwsUnaryEndpoint(awsEnv, signer, amendEndpoint(endpoint), codecAPI)
+      ): AwsUnaryEndpoint[F, Op, I, E, O, SI, SO] = {
+        val codecAPI = new AwsQueryCodecAPI(endpoint.id.name, service.version)
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/aws/internals/AwsQueryRPCInterpreter.scala:60]: Much simpler :+1:

**Comment by igor-ramazanov** [comment]
@Baccata It works! I've successfully got a response with an information from AWS CloudWatch.

File: modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala [hunk 2]
```diff
@@ -16,47 +16,133 @@
...
+          .getBytes("UTF-8")
+
+      case Right(_) =>
+        throw new IllegalStateException(
+          "Invalid codec: got XML decoder, must be AWS query encoder"
+        )
+    }
+}
+
+object AwsQueryCodecAPI {
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala]:
```suggestion
private[aws] object AwsQueryCodecAPI {
```

File: modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala [hunk 3]
```diff
@@ -16,47 +16,133 @@
...
+
+  /**
+    * Amend the schema to be able to work the the response xml payload, which wraps the output
+    * in two different layers of xml nodes.
+    *
+    * See https://smithy.io/2.0/aws/protocols/aws-query-protocol.html?highlight=aws%20query%20protocol#response-serialization
+    */
+  def xmlResponseSchema[A](
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala]:
```suggestion
  private def xmlResponseSchema[A](
```

File: modules/aws/src/smithy4s/query/AwsQueryEnrichment.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package smithy4s
+package aws.query
+
+import smithy4s.schema.Schema
+import smithy4s.schema.Schema.struct
+import smithy4s.schema.Schema.string
+
+private[aws] final case class AwsQueryEnrichment(
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/query/AwsQueryEnrichment.scala]: Unused class

File: modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala [hunk 5]
```diff
@@ -22,20 +22,22 @@ import smithy.api.{XmlFlattened, XmlName}
 import smithy4s.{Schema => _, _}
 
 private[aws] class AwsSchemaVisitorAwsQueryCodec(
-    val cache: CompilationCache[AwsQueryCodec]
+    val cache: CompilationCache[AwsQueryCodec],
+    val operationName: String,
+    val version: String
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/query/AwsSchemaVisitorAwsQueryCodec.scala]:
```suggestion
```

File: modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala [hunk 6]
```diff
@@ -337,7 +337,7 @@ object AwsQueryCodecSpec extends SimpleIOSuite {
     val cache: CompilationCache[AwsQueryCodec] =
       CompilationCache.make[AwsQueryCodec]
     val schemaVisitor: AwsSchemaVisitorAwsQueryCodec =
-      new AwsSchemaVisitorAwsQueryCodec(cache)
+      new AwsSchemaVisitorAwsQueryCodec(cache, operationName = "", version = "")
```
Comment (igor-ramazanov) [inline modules/aws/test/src/smithy4s/aws/query/AwsQueryCodecSpec.scala]:
```suggestion
      new AwsSchemaVisitorAwsQueryCodec(cache)
```

File: modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala [hunk 7]
```diff
@@ -16,47 +16,133 @@
...
-  ): Codec[A] = schema.compile(new AwsSchemaVisitorAwsQueryCodec(cache))
+  ): Codec[A] =
+    schema.hints.get(InputOutput) match {
+      case Some(InputOutput.Input) =>
+        val visitor =
+          new AwsSchemaVisitorAwsQueryCodec(
+            cache,
+            operationName,
+            serviceVersion
+          )
```
Comment (igor-ramazanov) [inline modules/aws/src/smithy4s/query/AwsQueryCodecAPI.scala]:
```suggestion
        val visitor = new AwsSchemaVisitorAwsQueryCodec(cache)
```

**Review by igor-ramazanov — CHANGES_REQUESTED** [review]
Removing/cleaning up a bit

**Comment by igor-ramazanov** [comment]
@Baccata Should we do anything else to support all of the [supported traits](https://smithy.io/2.0/aws/protocols/aws-query-protocol.html#supported-traits)?

**Comment by Baccata** [comment]
@igor-ramazanov yeah I believe we're gonna need to support a few of those at the very list. It'd be nice to run a search on the json specs available in the sdk repo to understand which of those are actually used, how often and in which context ? 

I'll add todo points to the comment in the feature branch

**Review by igor-ramazanov — APPROVED** [review]

**Pull request merged**